src/foam/nanos/ruler/Rule.js

```diff
@@ -230,6 +234,48 @@
           ruler.getDelegate().cmd_(x.put("OBJ", obj), getCmd());
         }
       `
+    },
+    {
+      name: 'reverseAction',
+      args: [
+        {
+          name: 'obj',
+          type: 'foam.core.FObject'
+        },
+        {
+          name: 'x',
+          type: 'foam.core.X'
+        }
+      ],
+      documentation: 'if one of the rules in a group throws an exception we need a way to reverse actions for previously executed rules.',
+      javaCode: ` `
+    },
+    {
+      name: 'updateRule',
+      type: 'foam.nanos.ruler.Rule',
+      documentation: 'since rules are stored as lists in the RulerDAO we use listeners to update them whenever ruleDAO is updated.' +
```

Comment:
Good idea.

Reply:
It would be simpler if we fetch relevant rules when applying so that we don't have to listen to changes and update the rule.

Reply:
Simpler but slower.